### PR TITLE
Non docker

### LINKED ISSUE #30: ssh builder
ssh implementation of builder, this can then be used in the vagrant implementation

### PR DESCRIPTION
**Pull request by shankj3**
...sorry for this  
closes #30 
Implemented ssh builder type and exec builder type. 
I also implemented some apple developer keychain credential stuff, but I didn't end up using it, as rod is already all logged in and i was at my wit's end with apple. I'm keeping it in here because i think that if i put waay more effort in I could get this working in a super secure way and it would alleviate everybodys' keychain pain. Idk though; likely wishful thinking.

---
## Changelist
- added new targets in makefile, dev-templates and darwin-werker
- added the concept of an ocelot prefix so files will not need to be downloaded to root 
- added `Init`, `Close` and `SetGlobalEnv` methods to the Builder Interface so that env's can be set on the makeItSo side of the code and not re-implemented in every builder implementation. `Init` ended up being for initiating an ssh connection, as well as `Close`... not sure if this was the best way @tjtelan as ssh ended up being the only one that needed it. 
- i wrote a _lot_ of tests. tests everywhere.
- started a vagrant implementation, but didn't finish it. it might be worth deleting, idk 
- for integrations, i switched to a pattern where the generated string is just saved to an environment variable then echoed in to the container or whatever
- developer profile integration
- split up `makeitso.go` into different files so its less of a behemoth. now there is also:
    - `prefight.go` --> runs `downloadCodebase`, `downloadBinaries`, `doIntegrations`
    - `prefightintegrate.go` --> home of `doIntegrations` and `downloadBinaries`
    - `stagerunner.go` --> runs the actual stages in the build config, handles failures and triggers
- added concept of tagging, so now either `buildImage` or `machineTag` is a required field. a werker will start up with a list of tags, and that will generate new queues. buildconf's that have the `machineTag` field will be put on these special queues.
- removed build stage required in validation

### CODE REVIEW

File: build/valet/contextvalet.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package valet
```
Comment (shankj3) [inline build/valet/contextvalet.go]: contextValet used to be in werker stream code kinda scattered around

File: build/streamer/werker_stream.go [hunk 2]
```diff
@@ -145,27 +145,4 @@ func (sp *StreamPack) ListenTransport(transpo chan *models.Transport) {
 		ocelog.Log().Debugf("adding info channel for hash %s to map for streaming access.", i.Hash)
 		go sp.processTransport(i)
 	}
-}
-
-func (sp *StreamPack) ListenBuilds(buildsChan chan *models.BuildContext, mapLock sync.Mutex) {
```
Comment (shankj3) [inline build/streamer/werker_stream.go]: this code wasn't being used, so i took it out

File: common/helpers/sshhelper/contextconnect.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+	"github.com/shankj3/ocelot/models"
+	"golang.org/x/crypto/ssh"
+)
+
+func getClient(facts *models.SSHFacts) (*ssh.Client, error) {
+	var auth []ssh.AuthMethod
+	switch {
+	case facts.Password != "":
+		auth = append(auth, ssh.Password(facts.Password))
+	case facts.KeyFP != "":
```
Comment (tjtelan) [inline common/helpers/sshhelper/contextconnect.go]: what does FP stand for?

File: common/helpers/sshhelper/contextconnect.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+	if scanner.Err() != nil {
+		logout <- []byte("An error has occured!")
+		logout <- []byte(scanner.Err().Error())
+	}
+}
+
+// RunAndLog runs a given command remotely via the ContextConnection's ssh client. The stdout and stderr will be processed
+// by the given StreamingFunc function and written to logout. the function will wait for the StreamingFunc function to
+// close its done channel on both the stdout processing and the stderr processing.
+func (c *Channel) RunAndLog(cmd string, envs []string, logout chan []byte,  streamingFunc StreamingFunc) error {
```
Comment (tjtelan) [inline common/helpers/sshhelper/contextconnect.go]: A part of me is wondering what signals the ssh client can support? (so, probably not SIGKILL, but SIGTERM, SIGINT, SIGQUIT, SIGUSR1, etc)... And what, if any, different kind of behaviors we'd want to support if we catch multiple signals.

File: build/builder/ssh/sshbuilder.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	} else {
+		res.Messages = []string{"Successfully established ssh connection " + models.CHECKMARK}
+	}
+	return res
+}
+
+// Setup for the BARE werker type will send off the checkout hash as the "docker id" on the docker id channel
+func (h *SSH) Setup(ctx context.Context, logout chan []byte, dockerIdChan chan string, werk *pb.WerkerTask, rc cred.CVRemoteConfig, werkerPort string) (*pb.Result, string) {
+	log.Log().Infof("setting up hash %s", werk.CheckoutHash)
+	dockerIdChan <- werk.CheckoutHash
```
[the remark this answers is not in the document]
Comment (shankj3) [reply build/builder/ssh/sshbuilder.go]: #36 just made this

**Pull request merged**